### ISSUE TITLE
Avoidable re-render when wrapperProps change triggers a dispatch down the tree

### ISSUE THREAD

**Issue by Hypnosphi** [reduxjs/react-redux#1354]
**Do you want to request a _feature_ or report a _bug_?**

A bug

**What is the current behavior?**

When the wrapped component dispatches any action in `cDU` or `uLE` as a reaction to external prop change, an unnecessary re-render is made

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to a CodeSandbox (https://codesandbox.io/s/new) or RN Snack (https://snack.expo.io/) example below:**

https://codesandbox.io/s/unruffled-glitter-ofxmq?fontsize=14

Open console, click "Activate"

**What is the expected behavior?**

`Child` re-renders only once after clicking the button

**Which versions of React, ReactDOM/React Native, Redux, and React Redux are you using?  Which browser and OS are affected by this issue? Did this work in previous versions of React Redux?**

See codesandbox above

The probable reason of the bug is that `lastWrapperProps.current` is outdated at the moment of dispatch: 
https://github.com/reduxjs/react-redux/blob/master/src/components/connectAdvanced.js#L322-L325
This leads to `ownProps` reference being reset to an older value:
https://github.com/reduxjs/react-redux/blob/master/src/connect/selectorFactory.js#L78

One possible workaround is to introduce a separate `recentWrapperProps` ref and set in immediately inside `render`

```
const recentWrapperProps = useRef()
recentWrapperProps.current = wrapperProps
```

**Comment by ckedar** [comment]
The issue is indeed because of stale `lastWrapperProps`.

At present the value of `lastWrapperProps` is being updated in `uLE`. 
It should ideally be done during the render itself. 

So moving the assignment `lastWrapperProps.current = wrapperProps` from `uLE` to just before return resolves the issue while passing all test cases.

Updating  of other two  ref values - `lastChildProps` and `renderIsScheduled` should also be moved likewise.

**Comment by markerikson** [comment]
@ckedar : no, we should _not_ modify refs while rendering.  That should only ever be doing in the commit phase, once we know the component has finished rendering.

**Comment by markerikson** [comment]
I'm going to close this as a WONTFIX for now, largely because we're encouraging folks to stop using `connect` (even though it'll be fully supported in v8).

If someone still feels strongly about this please comment, or even better file a PR to improve behavior here.
